src/main/scala/services/ValidationService.scala

```diff
@@ -1,11 +1,71 @@
 package com.fortysevendeg.tagless.sample
 package services
 
-import cats.data.ValidatedNel
+import cats.data.{ValidatedNec, ValidatedNel}
+import cats.effect.Sync
+import cats.implicits._
 import com.fortysevendeg.tagless.sample.models._
+import org.slf4j.{Logger, LoggerFactory}
 
 trait ValidationService[F[_]] {
 
   def validateUser(name: String, age: Int, gender: Char, heightAndWeight: String): F[ValidatedNel[ValidationError, UserInformation]]
 
 }
+
+object ValidationService {
+
+  type ValidationResult[A] = ValidatedNec[ValidationError, A]
+
+  private def toGender(gender: Char): Either[ValidationError, Gender] = gender match {
+    case 'M' => Either.right(Male)
+    case 'F' => Either.right(Female)
+    case _ => Either.left(InvalidGenderError)
+  }
+
+  def validateAgeLogged[F[_]: Sync](age: Int)(logger: Logger): F[ValidationResult[Int]] = Sync[F].delay {
+    Either.cond[ValidationError, Int](age >= 18, age, NotAdultError)
+      .toValidatedNec
+  }.handleErrorWith { _ =>
```

Comment:
this `handleErrorWith` is capturing errors, but we're not raising anyone, so it's never called (that my guess, you're not seeing this logs in the console when executing the test right?).

We also do not need the `delay` here, since this is not a "side effect". We need to find a way to transform the `Either[ValidationError, Int]` (I love you approach of `Either.cond`) to `F[ValidationResult[Int]]` logging in the case of a `Left` but keeping that value.

Aside note, you're raising an `EncryptionError` (but this code never gets executed by the above)

Reply:
Thanks for the review!!

I'm kind of confused about what you mention about `handleErrorWith`. As you say I can't see the logs. I was under the impression that if each validation is under F (I'll change that later) and each one raise an error, by flatMapping through all of them and then performing the `mapN` operation, I would still be raising the error... but now that I'm thinking about it, it doesn't add up.

Should I leverage that `raiseError` to the final step in `validateUser`?

Reply:
I wouldn't raise any errors. I think it's my fault because I showed each validation def typed to `F` but the only reason for that is for adding a log statement in case of left.

My solution looks like the following:

```scala
def log(): F[Unit] = ???
val either: Either[Error, Int] = ???
either.fold(e => log().map(_ => InvalidNec(e)), i => Sync[F].pure(Valid(I)))
```

Didn't compiled, just a draft. Let me know if makes sense